HospitalSystem.Core/Appointments/AppointmentService.cs

```diff
@@ -0,0 +1,399 @@
+            {
+                return false;
+            } 
+        }
+        return true;
+    }
+
+    public int CompareCheckupsByDoctorsName(Checkup checkup1, Checkup checkup2)
+    {
+        string name1 = _doctorService.GetById((ObjectId)checkup1.Doctor.Id).FirstName;
+        string name2 = _doctorService.GetById((ObjectId)checkup2.Doctor.Id).FirstName;
+        return String.Compare(name1, name2);
+    }
+
+    public int CompareCheckupsByDoctorsSpecialty(Checkup checkup1, Checkup checkup2)
+    {
+        string specialty1 = _doctorService.GetById((ObjectId)checkup1.Doctor.Id).Specialty.ToString();
+        string specialty2 = _doctorService.GetById((ObjectId)checkup2.Doctor.Id).Specialty.ToString();
+        return String.Compare(specialty1, specialty2);
+    }
+
+    public List<Checkup> GetEarliestFreeCheckups(DateRange interval, Specialty speciality, int numberOfCheckups, User user)
+    {
+        List<Checkup> checkups = new List<Checkup>();
+        DateTime iterationDate = RoundUp(DateTime.Now,TimeSpan.FromMinutes(15));
+
+        while ( checkups.Count < numberOfCheckups)
+        { 
+            if (iterationDate.TimeOfDay >=Globals.ClosingTime.TimeOfDay)
+            {
+                iterationDate = new DateTime(iterationDate.Year, iterationDate.Month, iterationDate.Day, Globals.OpeningTime.Hour, Globals.OpeningTime.Minute, Globals.OpeningTime.Second);
+                iterationDate = iterationDate.AddDays(1);
+                continue;
+            }
+
+            if (iterationDate.TimeOfDay < Globals.OpeningTime.TimeOfDay)
+            {
+                iterationDate = new DateTime(iterationDate.Year, iterationDate.Month, iterationDate.Day, Globals.OpeningTime.Hour, Globals.OpeningTime.Minute, Globals.OpeningTime.Second);
+                continue;
+            }
+
+            if (!(interval.Starts.TimeOfDay<=iterationDate.TimeOfDay && iterationDate.TimeOfDay<interval.Ends.TimeOfDay))
+            {
+                iterationDate = iterationDate.AddMinutes(15);
+                continue;
+            }
+
+            foreach (Doctor doctor in _doctorService.GetManyBySpecialty(speciality))
+            {
+                Checkup newCheckup = new Checkup(
+                    iterationDate,
+                    new MongoDB.Driver.MongoDBRef("patients", user.Person.Id),
+                    new MongoDB.Driver.MongoDBRef("doctors", doctor.Id),
+                    "no anamnesis");
+                if (!IsDoctorAvailable(newCheckup.DateRange, doctor))
+                {
+                    continue;
+                }
+                else
+                {
+                    if (checkups.Count >= numberOfCheckups)
+                    {
+                        break;
+                    }
+
+                    checkups.Add(newCheckup);
+                }
+            }
+            iterationDate = iterationDate.AddMinutes(15);
+        }
+        return checkups;
```

Comment:
Move a lot of this logic to a smart TimeSlot class. E.g. it has ts.Next() which gives next